core/src/main/java/com/linecorp/armeria/common/RequestContext.java

```diff
@@ -435,6 +435,26 @@ default ByteBufAllocator alloc() {
     @MustBeClosed
     SafeCloseable push();
 
+    /**
+     * Immediately run a given {@link Runnable} with this context.
+     */
+    default void run(Runnable runnable) {
+        requireNonNull(runnable, "runnable");
+        try (SafeCloseable ignored = push()) {
+            runnable.run();
+        }
+    }
+
+    /**
+     * Immediately call a given {@link Callable} with this context.
+     */
+    default <T> T run(Callable<T> callable) throws Exception {
```

Comment:
I guess `Supplier` is more commonly used for Java 8+.
Is it better to support `Supplier<T>` than `Callable<T>` becase the minium requirement of Java 8 for Armeria. 🤔

Reply:
How about offering both `Supplier<T>` and `Callable<T>`?

Reply:
If we support both `Supplier` and `Callable`, users can not use a lambda expression.
For example, Java compiler can not compile `ctx.run(() -> ...)` because it is an ambiguous method call.
Because of the limitation, we need to choose one between `Supplier` and `Callable`.

Reply:
#### [Callable](https://docs.oracle.com/javase/8/docs/api/index.html?java/util/concurrent/Callable.html)
> The Callable interface is similar to Runnable, in that both are designed for classes whose instances are potentially **executed by another thread**. A Runnable, however, does not return a result and cannot throw a checked exception.

#### [Supplier](https://docs.oracle.com/javase/8/docs/api/java/util/function/Supplier.html)
> Represents a supplier of results. There is no requirement that a new or distinct result be returned each time the supplier is invoked.

`Callable` is also designed to be run on another thread. on the other hand, `Supplier` is very general. It just "supplies a value" and that's it.

Shouldn't we run a given callable or supplier in another thread using the current context?
If so, I think callable looks better.
What do you think?

Reply:
Callable seems better - especially since we're focused on lambda usage, Callable seems to just be a superset of Supplier

Reply:
> Shouldn't we run a given callable or supplier in another thread using the current context?

That could be right or wrong. A single event loop is shared by multiple requests. So a thread that initiates a request could not have a context in the thread-local.

>  If so, I think callable looks better.

I realized that `makeContextAware` already uses `Callable<T>`. Let's use `Callable` for consistency. 👍
https://github.com/line/armeria/blob/97e23218b95f6e047d29ac75cf6ab53ede1ea4a5/core/src/main/java/com/linecorp/armeria/common/RequestContext.java#L485

Reply:
`throws Exception` may be very annoying though because a user will always have to wrap it with `try-catch`. :thinking:

Reply:
From that point of view, the `supplier` looks better than `callable`. 🤔

Reply:
But `Callable` is better in that it allows a user to throw anything inside the `Callable`. :laughing: Let's leave this as it is for now..